src/completion.ml

```diff
@@ -0,0 +1,228 @@
+(** Completion procedure for closed equations *)
+
+open Terms
+open Basics
+open Extra
+open Timed
+
+(** [lpo ord] computes the lexicographic path ordering corresponding to
+    the strict total order [ord] on the set of all function symbols. *)
+let rec lpo : sym Ord.cmp -> term Ord.cmp = fun ord t1 t2 ->
+  let f, args = get_args t1 in
+  let f = get_symb f in
+  match f with
+  | None   -> if t1 == t2 then 0 else -1
+  | Some f ->
+      if List.exists (fun x -> lpo ord x t2 >= 0) args then 1
+      else
+        let g, args' = get_args t2 in
+        let g = get_symb g in
+        match g with
+        | None   -> 1
+        | Some g ->
+            match ord f g with
+            | k when k > 0 ->
+                if List.for_all (fun x -> lpo ord t1 x > 0) args' then 1
+                else -1
+            | 0            ->
+                if List.for_all (fun x -> lpo ord t1 x > 0) args' then
+                  Ord.ord_lex (lpo ord) args args'
+                else -1
+            | _            -> -1
+
+(** [to_rule (lhs, rhs)] tranlates the pair [(lhs, rhs)] of closed terms into
+    a rule. *)
+let to_rule : term * term -> sym * rule = fun (lhs, rhs) ->
+  let (s, args) = get_args lhs in
+  let s = get_symb s in
+  match s with
+  | None   -> assert false
+  | Some s ->
+      let vs = Bindlib.new_mvar te_mkfree [||] in
+      let rhs = Bindlib.unbox (Bindlib.bind_mvar vs (lift rhs)) in
+      s, { lhs = args ; rhs = rhs ; arity = List.length args ; vars = [||] }
+
+(** [find_deps eqs] returns a pair [(symbs, deps)], where [symbs] is a list of
+    (names of) new symbols and [deps] is a list of pairs of symbols. (s, t)
+    is added into [deps] iff there is a equation (l, r) in [eqs] such that
+    (l = s and t is a proper subterm of r) or (r = s and t is a proper subterm
+    of l). In this case, we also add s and t into [symbs]. *)
+let find_deps : (term * term) list -> string list * (string * string) list
```

Comment:
This seems to be related to the computation of dependency pairs. Check with @GuillaumeGen whether you could factorize/share some code.

Reply:
Complexity of the algorithm?